### DISCUSSION TITLE
Experiment setup with a custom model

### DISCUSSION THREAD

**Discussion by isaacbasil** [pybamm-team/PyBaMM#4503] [Q&A]
I am having some trouble running experiments in Pybamm with a custom model (which is basically the DFN). My goal is to discharge an NMC half-cell until a voltage constraint is reached, followed by a relaxation period for a prescribed length of time. Also, I need to be able to provide t_eval. When I try the following, 
```
        discharge_step = pybamm.step.current(current_value, termination="2.7 V")
        relaxation_step = pybamm.step.current(0, duration="200 seconds")
        experiment = pybamm.Experiment([discharge_step, relaxation_step])
        self.simulation = pybamm.Simulation(self.model, experiment=experiment)
```

I get the following error:
"Cannot update parameter 'Current function [A]' as it does not have a default value. ('Current function [A]' not found. Best matches are []). If you are sure you want to update this parameter, use param.update({{name: value}}, check_already_exists=False)"

However, I am not sure how to set a default value for 'Current function [A]'. Or alternatively, I am not sure where to place param.update({{name: value}}, check_already_exists=False). 

Any advice appreciated :)

**Comment by martinjrobins** [comment]
does your model have a `default_parameter_values` property set, if you don't provide one to `pybamm.Simulation` this is what it uses as the parameter values object. Also, does your model have a `Current function [A]` parameter in it somewhere?

If you are able to provide a minimal script that can reproduce the error this would be helpful

**Reply by isaacbasil** [reply]
I added a parameter_values argument when I call pybamm.Simulation and that fixed the above problem. I then run into some other issues, but the main problem which I'm not sure if I'll be able to get around is passing the t_eval argument, since it looks like the experiment feature always overrides this? Also, I'm starting to think it doesn't make sense for me to use the Simulation feature, since I'm using a custom model and so the model is processed step by step. Is it possible to do an experiment without using the Simulation class? 

I'll work on getting a minimal script together.

**Reply by rtimms** [reply]
For a model to work with experiments it needs to have

1. All the Variable objects in the variables dict
2. The "Current function [A]" and "Battery voltage [V]" variables

See https://github.com/pybamm-team/PyBaMM/pull/3995

**Reply by rtimms** [reply]
Why do you need to provide `t_eval` as well as an experiment? You can interpolate the solution onto any times you like.

**Reply by isaacbasil** [reply]
Thanks for the comments. That's a good point, interpolating would get around the t_eval issue. Is it possible to use the experiment feature without using the Simulation class? Since I am processing the model step by step, as in [this](https://docs.pybamm.org/en/stable/source/examples/notebooks/simulations_and_experiments/simulation-class.html) tutorial, and I guess calling the simulation class would repeat these steps?

**Reply by rtimms** [reply]
You need to use a simulation to use an experiment, but you should be able to just pass your model, mesh etc. to the simulation instead of processing the model step by step.

**Reply by isaacbasil** [reply]
ok cheers, I'll come back to this on Monday.

**Reply by isaacbasil** [reply]
I was already half way through doing this so I thought I'd try it - I solved the model for the discharge until 2.7V as normal, and then solved the model again using set_initial_conditions_from(previous_solution) for the relaxation period. That seems to work for now. Perhaps I'll switch to doing it with the Simulation class in the future. Thanks for your help!
